### PR TITLE
[Flax] Add finetune Stable Diffusion

### PR DESCRIPTION
**Pull request by duongna21**
# What does this PR do?

- Add Flax example for finetuning Stable Diffusion. 
- Ran well on Tesla A100 (40GB) with max batch size = 3. 
- EMA is not added because my 1-line tree_map implementation for EMA make training much slower, and it did not show a visible improvement on the result.  

How to run (76% faster than PyTorch example with same args on Tesla A100)
```
export MODEL_NAME="duongna/stable-diffusion-v1-4-flax"
export dataset_name="lambdalabs/pokemon-blip-captions"

python train_text_to_image_flax.py \
  --pretrained_model_name_or_path=$MODEL_NAME \
  --dataset_name=$dataset_name \
  --resolution=512 --center_crop --random_flip \
  --train_batch_size=1 \
  --max_train_steps=15000 \
  --learning_rate=1e-05 \
  --max_grad_norm=1 \
  --output_dir="sd-pokemon-model" 
```
Prompt: `robotic cat with wings`

ảnh

## Who can review?
cc @patrickvonplaten @patil-suraj

### CODE REVIEW

**Comment by patil-suraj** [comment]
Very cool PR @duongna21 , amazing, will take a look soon !

File: examples/text_to_image/train_text_to_image_flax.py [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+        }
+        batch = {k: v.numpy() for k, v in batch.items()}
+
+        return batch
+
+    total_train_batch_size = args.train_batch_size * jax.local_device_count()
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset, shuffle=True, collate_fn=collate_fn, batch_size=total_train_batch_size
+    )
+
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image_flax.py:362]: Let's drop the last incomplete batch here, as incomplete batches can't be sharded on multi-device 
`drop_last=True`

File: examples/text_to_image/train_text_to_image_flax.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+    # Load models and create wrapper for stable diffusion
+    tokenizer = CLIPTokenizer.from_pretrained(args.pretrained_model_name_or_path, subfolder="tokenizer")
+    text_encoder = FlaxCLIPTextModel.from_pretrained("openai/clip-vit-large-patch14")
+    vae, vae_params = FlaxAutoencoderKL.from_pretrained(args.pretrained_model_name_or_path, subfolder="vae")
+    unet, unet_params = FlaxUNet2DConditionModel.from_pretrained(args.pretrained_model_name_or_path, subfolder="unet")
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image_flax.py:367]: Let's pass `dtype` here to allow mixed-precision training.

**Review by patil-suraj — COMMENTED** [review]
Looks very good, thanks a lot for adding this example. Left some comments about dataloader and `dtype`. 

Also, let's update the readme with an example command to show how to run these examples.

**Comment by duongna21** [comment]
@patil-suraj Thanks for very helpful comments. Addressed them!

**Review by patil-suraj — APPROVED** [review]
Thanks for addressing the comments, good for merge now!

**Comment by entrpn** [comment]
@duongna21 Thank you for this contribution. I did have to make a change in order to get it working with TPUs.

```python
device_type = jax.devices()[0].device_kind

weight_dtype = torch.float32
if 'TPU' in device_type:
    weight_dtype = jnp.float32
    if args.mixed_precision == "fp16":
        weight_dtype = jnp.float16
    if args.mixed_precision == "bf16":
        weight_dtype = jnp.bfloat16
else:
    if args.mixed_precision == "fp16":
        weight_dtype = torch.float16
    elif args.mixed_precision == "bf16":
        weight_dtype = torch.bfloat16
```

**Comment by duongna21** [comment]
@entrpn Thanks a lot! Will be fixed at #1038.

**Pull request merged**